cachelib/navy/common/FdpNvme.cpp

```diff
@@ -0,0 +1,228 @@
+  if (cmd == nullptr) {
+    throw std::invalid_argument("Uring cmd is NULL!");
+  }
+  memset(cmd, 0, sizeof(struct nvme_uring_cmd));
+  cmd->opcode = opcode;
+
+  uint64_t sLba = start >> nvmeData.lbaShift();
+  uint32_t nLb  = (size >> nvmeData.lbaShift()) - 1;
+
+  /* cdw10 and cdw11 represent starting lba */
+  cmd->cdw10 = sLba & 0xffffffff;
+  cmd->cdw11 = sLba >> 32;
+  /* cdw12 represent number of lba's for read/write */
+  cmd->cdw12 = (dtype & 0xFF) << 20 | nLb;
+  cmd->cdw13 = (dspec << 16);
+  cmd->addr = (uint64_t)buf;
+  cmd->data_len = size;
+
+  cmd->nsid = nvmeData.nsId();
+}
+
+void FdpNvme::prepReadUringCmdSqe(struct io_uring_sqe& sqe, int fd, void* buf,
+    size_t size, off_t start) {
+  // Placement Handle is not used for read.
+  prepFdpUringCmdSqe(sqe, fd, buf, size, start, nvme_cmd_read, 0, 0, getNvmeData());
+}
+
+void FdpNvme::prepWriteUringCmdSqe(struct io_uring_sqe& sqe, int fd, void* buf,
+    size_t size, off_t start, int handle) {
+  static constexpr uint8_t kPlacementMode = 2;
+  uint16_t id;
+
+  if (handle == -1) {
+    id = 0; // Use the default stream
+  } else if (handle >= 0) {
+    id = static_cast<uint16_t>(handle);
+  } else {
+    throw std::invalid_argument("Invalid placement identifier!");
+  }
+
+  prepFdpUringCmdSqe(sqe, fd, buf, size, start, nvme_cmd_write,
+              kPlacementMode, id, getNvmeData());
+}
+
+NvmeData FdpNvme::readNvmeInfo(int fd) {
+  int namespace_id = ioctl(fd, NVME_IOCTL_ID);
+  if (namespace_id < 0) {
+    XLOG(ERR)<< "failed to fetch namespace-id, fd "<< fd;
+    return NvmeData{};
+  }
+
+  char ctrl[NVME_IDENTIFY_DATA_SIZE]; // Identify ctrl data
+  struct nvme_passthru_cmd cmd_ctrl = {
+    .opcode         = nvme_admin_identify,
+    .nsid           = 0,
+    .addr           = (uint64_t)(uintptr_t)ctrl,
+    .data_len       = NVME_IDENTIFY_DATA_SIZE,
+    .cdw10          = NVME_IDENTIFY_CNS_CTRL,
+    .cdw11          = NVME_CSI_NVM << NVME_IDENTIFY_CSI_SHIFT,
+    .timeout_ms     = NVME_DEFAULT_IOCTL_TIMEOUT,
+  };
+
+  int err = ioctl(fd, NVME_IOCTL_ADMIN_CMD, &cmd_ctrl);
+  if (err) {
+    XLOG(ERR)<< "failed to fetch identify ctrl";
+    return NvmeData{};
+  }
+
+  static constexpr uint16_t kMDTSOffset = 77u;
+  uint8_t mdts = (uint8_t)ctrl[kMDTSOffset];
+  uint32_t maxTfrSize = (1 << mdts) * getpagesize();
+
+  struct nvme_id_ns ns;
+  struct nvme_passthru_cmd cmd = {
+    .opcode         = nvme_admin_identify,
+    .nsid           = (uint32_t)namespace_id,
+    .addr           = (uint64_t)(uintptr_t)&ns,
+    .data_len       = NVME_IDENTIFY_DATA_SIZE,
+    .cdw10          = NVME_IDENTIFY_CNS_NS,
+    .cdw11          = NVME_CSI_NVM << NVME_IDENTIFY_CSI_SHIFT,
+    .timeout_ms     = NVME_DEFAULT_IOCTL_TIMEOUT,
+  };
+
+  err = ioctl(fd, NVME_IOCTL_ADMIN_CMD, &cmd);
+  if (err) {
+    XLOG(ERR)<< "failed to fetch identify namespace";
+    return NvmeData{};
+  }
+
+  auto lbaShift = (uint32_t)ilog2(1 << ns.lbaf[(ns.flbas & 0x0f)].ds);
```

Comment:
The `ds` should be in terms of power of two already?

Reply:
yes, you are right. will correct it